### PR TITLE
Support podman

### PR DESCRIPTION
**Pull request by tmds**
This makes tye work with rootless podman.

It requires podman 3.1+.

Users need to update their `$HOME/.config/containers/containers.conf` to allow containers to access services on the host:
```
[engine]

network_cmd_options=["allow_host_loopback=true"]
```

All tests pass except `NginxIngressTest` which fails for me due to https://github.com/dotnet/tye/issues/428.

Fixes https://github.com/dotnet/tye/issues/563

@davidfowl @jkotalik ptal

### CODE REVIEW

**Comment by tmds** [comment]
I ran these tests in a clean Fedora 33 VM.

I did the following configuration:
```sh
mkdir -p $HOME/.config/containers
cat >$HOME/.config/containers/containers.conf <<EOF
[engine]

network_cmd_options=["allow_host_loopback=true"]
EOF

sudo dnf -y update podman
```

Then ran:
```sh
./build.sh --test
```

All tests passed. For E2E:

```
 Assemblies Run
Microsoft.Tye.E2ETest.dll
Summary
Tests run: 81 — Skipped: 1, Run time: 1213.693s, Finished: 04/22/2021 09:44:38

Skipped tests
Skipped ❢  E2ETest.TyeRunTests.FrontendBackendAzureFunctionTest
```

File: src/Microsoft.Tye.Core/DockerDetector.cs [hunk 1]
```diff
@@ -14,41 +18,107 @@ public class DockerDetector
...
+            AspNetUrlsHost = "localhost";
             try
             {
-                var result = await ProcessUtil.RunAsync("docker", "version", throwOnError: false, cancellationToken: new CancellationTokenSource(Timeout).Token);
-                return result.ExitCode == 0;
+                ProcessResult result;
+                try
+                {
+                    // try to use podman.
+                    result = ProcessUtil.RunAsync("podman", "version -f \"{{ .Client.Version }}\"", throwOnError: false, cancellationToken: new CancellationTokenSource(Timeout).Token).Result;
```
Comment (jkotalik) [inline src/Microsoft.Tye.Core/DockerDetector.cs:43]: It's been a while, but is it possible for someone to have both podman and docker locally? If so, having a way to select which one should be used?

Comment (tmds) [reply src/Microsoft.Tye.Core/DockerDetector.cs:43]: Yes, both can be installed side-by-side. In most cases, only one will be installed.
Do you want to make it configurable in `tye.yaml`, or in some other way?

**Comment by tmds** [comment]
@jkotalik can you review the PR? Functionally I think it's fine, but you may want to suggest some refactorings. I hope this can be part of the next tye release.

File: src/Microsoft.Tye.Core/DockerDetector.cs (hunk 1 above)
Comment (jkotalik) [reply src/Microsoft.Tye.Core/DockerDetector.cs:43]: I think it needs to be configurable.

**Comment by jkotalik** [comment]
LGTM besides an option to turn on and off.

**Comment by tmds** [comment]
@jkotalik I've added an option to `tye.yaml` that allows to select `podman`/`docker`. ptal at the changes.

File: src/schema/tye-schema.json [hunk 2]
```diff
@@ -13,6 +13,11 @@
             "description": "Dockerhub username or hostname of remote registry. Used for tagging images.",
             "type": "string"
         },
+        "engine": {
```
Comment (jkotalik) [inline src/schema/tye-schema.json:16]: I think containerEngine is a better name here.

**Review by jkotalik — APPROVED** [review]
Pending a rename.

**Comment by tmds** [comment]
I've renamed the property.

I also relaxed the podman version check because I don't want risk it making things unusable. Instead of failing when we can't parse the version, we assume the version requirement is met.

**Comment by jkotalik** [comment]
Thanks!

**Comment by tmds** [comment]
@jkotalik thank you too. Do you have a rough idea when you'll do the next release?

**Comment by jkotalik** [comment]
Hoping for one relatively soon 👍 . There are a few key bugs we need to fix first (.NET 6 support).

**Pull request merged**